### DISCUSSION TITLE
Code formatting

### DISCUSSION THREAD

**Discussion by brunetton** [jejellyroll-fr/fpdb-3#42] [development]
There are trailing white spaces everywhere in the project. Every decent code editor nowadays are removing those trailing spaces by default, that makes two mains problems:

- new contributors may commit white spaces changes in their commits if they're not using tools like [Git gui](https://git-scm.com/docs/git-gui/fr) to review their commits before commiting
- those white spaces makes a tons of false positive changes when trying to make [atomic commits](https://dev.to/samuelfaure/how-atomic-git-commits-dramatically-increased-my-productivity-and-will-increase-yours-too-4a84) and make commit process really slower and error-prone

![image](https://github.com/user-attachments/assets/38d85404-1079-4738-be9d-09505ea1fe26)

![image](https://github.com/user-attachments/assets/2d094d03-2de3-4586-9bc2-46d214d6e367)

## Proposed solution

I suggest to make use of [Ruff](https://github.com/astral-sh/ruff), that is a formatter (AND a linter, too), written in Rust, so super fast. The configuration is made [in the pyproject.toml file](https://docs.astral.sh/ruff/configuration/) so it'll be more obvious to newcomers that we use this tool on this project (of course it has to be mentioned in README/Wiki)

I propose to make a PR for a big cleaning commit removing trailing white-spaces and adding a Ruff config that seems to correspond to that way the code is on this project (I don't want to reformat existing code, or only marginally to prevent reformatting conflicts later)

What do you think about @jejellyroll-fr ?

**Comment by jejellyroll-fr** [comment]
good idea ;)

**Comment by brunetton** [comment]
Here we are: https://github.com/jejellyroll-fr/fpdb-3/pull/45

If (like me) you're using VScode, here's the configuration to use Ruff and format on save:

```json
  "[python]": {
    "editor.rulers": [120],
    "editor.defaultFormatter": "charliermarsh.ruff",
    "editor.formatOnSave": true
  }
```

I use [UV](https://docs.astral.sh/uv/) to manage Python tools. To install Ruff it with UV (in an isolated Python environnement, and available in $PATH):
```
uv tool install ruff
```

And to check the code and list errors:
- `ruff check .`
- `ruff check --statistics` for a summary

**Reply by jejellyroll-fr** [reply]
Perhaps Can be part of wiki 😉

Thanks a lot Bruno...really cool ☺️

**Reply by jejellyroll-fr** [reply]
@brunetton 
these ruff and uv tools are great, I've never really used them before ... I'm just discovering them:

<img width="573" alt="Screenshot 2024-09-29 at 13 39 40" src="https://github.com/user-attachments/assets/80b0f73d-af80-4010-a4a8-17e2ae69e20a">


so I've started to include it in the CI:
<img width="623" alt="Screenshot 2024-09-29 at 13 40 04" src="https://github.com/user-attachments/assets/cbfeda9b-439a-4dfb-b88a-5f372b611820">


I promise I'll try to make atomics commits 😉

**Reply by brunetton** [reply]
> Perhaps Can be part of wiki 😉
> 
> Thanks a lot Bruno...really cool ☺️

Of course :)

**Reply by brunetton** [reply]
> so I've started to include it in the CI:


Not sur the overall code quality is ready for so "high-level" quality checks 🤔 But yes ideally in a perfect world this could be part of CI. 
At least using it with code editor is a good start. I personally prefer focusing on user experience (there's a LOT to do) and global code structure (a LOT to do also I think) and not touching existing code a so much (that will probably be rewritten).

**Reply by brunetton** [reply]
> I promise I'll try to make atomics commits 😉

I'll be happy to see :) :+1:

**Comment by brunetton** [comment]
Hi @jejellyroll-fr 

Coming back to code after a while. I wanted to fix tests, and most of all add more testing around stats (because they are almost all false or 0 in my case). So I jumped in tests files, and my eyes exploded seeing a tons of trailing spaces :(

<img width="471" height="208" alt="image" src="https://github.com/user-attachments/assets/47ec3c52-282d-4e9f-b536-6a2bd71cf4a9" />

Didn't we agree to use **Ruff** for code formatting ? This file is clearly not formatted (at all, with any formater). It's a shame as again, this could lead in super complex conflicts between maintainers :'( 

I'll have to start by adding a purely formatting commit, which is quite a shame (shouldn't exists; all commits should be "real" changes, not formatting commits that makes the commit log harder to follow)

**Reply by brunetton** [reply]
Make sure you activated Ruff in VScode (if you're using VScode) and auto-format on save, so you won't have to worry about formatting, ever :)

https://github.com/jejellyroll-fr/fpdb-3/discussions/42#discussioncomment-10786368

**Reply by brunetton** [reply]
After a closer look; there are still **67 files** that contains trailing spaces:

```
./Hud.py
./PokerStarsToFpdb.py
./ThemeManager.py
./ConfigInitializer.py
./HandDataReporter.py
./test/test_pokerstars_showdown.py
./test/test_pokerstars_progressive_bounties.py
./test/test_pokerstars_readholecards.py
./test/test_hud_popup_tooltips.py
./test/test_pokerstars_markstreets.py
./test/test_pokerstars_bovada_adjustments.py
./test/test_hud_simple.py
./test/test_pokerstars_process_split_bounty.py
./test/test_rake_fixes_integration.py
./test/test_pokerstars_regular_bounties.py
./test/test_calc_check_call_raise.py
./test/test_guihandviewer_rake_display.py
./test/test_hud_interactive_manual.py
./test/test_pokerstars_collectpot.py
./test/test_pokerstars_error_cases.py
./test/test_theme_manager.py
./test/test_hud.py
./test/test_custom_themes.py
./test/test_card.py
./test/test_pokerstars_processaction.py
./test/test_pokerstars_cashout_rake.py
./test/test_BovadaToFpdb.py
./test/test_guihandviewer_uncalled_bet.py
./test/test_pokerstars_readaction.py
./test/test_popup_performance.py
./test/test_hud_stat_set_switching.py
./test/test_xml_theme_persistence.py
./test/test_theme_creator_dialog.py
./test/test_theme_config_integration.py
./test/test_popup_simple.py
./test/test_popup.py
./test/test_pokerstars_process_hand_info_key.py
./test/test_aux_simple.py
./test/test_cashout_fees_storage.py
./test/test_pokerstars_addcollectpot_adjustments.py
./test/test_pokerstars.py
./test/test_pokerstars_cashout.py
./test/test_pokerstars_readcommunity.py
./test/test_pokerstars_readbutton.py
./test/test_modern_popup_improvements.py
./test/test_pokerstars_processraise.py
./test/test_action_counting.py
./test/test_pokerstars_parse_rake_and_pot.py
./test/test_aggr.py
./test/test_card_comprehensive.py
./test/test_pokerstars_readplayerstacks.py
./test/test_config_initializer.py
./test/test_HUD_main.py
./test/test_hand_cashout_pot.py
./test/test_modern_popup.py
./test/test_cashout_fees_migration.py
./test/test_pokerstars_readhandinfo.py
./test/test_calc_cbets.py
./ThemeCreatorDialog.py
./Card.py
./regression-test-files/cash/TheBigGame/TBG-grabber.py
./Importer.py
./DerivedStats.py
./Hand.py
./ConfigObservers.py
./ConfigurationManager.py
./HudStatsPersistence.py
```

Do you want me to make another big cleaning commit like I did [here](https://github.com/jejellyroll-fr/fpdb-3/pull/45/commits/734f038ab1cf0b8602514d897c7b82d183258a21)

This is quite depressing to make it again as it's a very basic requirement for starting to work together on the same project

**Comment by brunetton** [comment]
Also, why enabling **ALL** Linting rules in this project ? 

https://github.com/jejellyroll-fr/fpdb-3/blob/6b7d4bc2f09623c8fb2ffd270d624340aba9212e/pyproject.toml#L76

I don't think we are in a state where the code is 99% perfect ! Far from it. Why don't we use standard configuration (except for line length), that don't alert for missing docstring  in public functions for example. 

And one day, when the codebase will be more stable and better quality, we can enable more and more checks.

https://docs.astral.sh/ruff/linter/#rule-selection

I propose to remove this `ALL` selection to let the standard one. Are you opposed to it ?

**Comment by jejellyroll-fr** [comment]
Thanks for your feedback. I hear your point about Ruff and trailing whitespace, and yes, we need to fix that to avoid unnecessary conflicts.

I suggest we simply resolve this with a global formatting commit and an appropriate Ruff configuration, then move on to more critical issues. This will prevent this kind of detail from taking up more space than necessary.

**Reply by brunetton** [reply]
Yes that's what I proposed:

> Do you want me to make another big cleaning commit like I did [here](https://github.com/jejellyroll-fr/fpdb-3/pull/45/commits/734f038ab1cf0b8602514d897c7b82d183258a21)

**Reply by brunetton** [reply]
BUT unfortunately there's no option to only take care of spaces (and it's logical, if we use Ruff, it's for all its functionalities and its style). So I started to manually selecting changes but it's really long.

The majority of reformatting Ruff does is to separate arguments of functions calls; like this:

<img width="828" height="161" alt="image" src="https://github.com/user-attachments/assets/307c50b1-721b-4ba7-87f1-eecf7921a353" />

- advantage: we clearly see each argument
- disadvantage: it takes some place, sometimes a lot of lines when arguments are "obvious" 

I've the feeling that you didn't adopted the "blinded" syntax formatting and that you like to format your code with a certain style (like calling function with arguments in the same line - I do like this too, but depends on cases and ideally I'd like the formatting tool to let me choose, that is not the case with Ruff). If this is the case maybe Ruff (as formatter) is too "advanced" for this project and we should go for something much simpler that only
- takes care of trailing spaces
- reformat lines that are too long (?)

Maybe we can decide to stop using Ruff and only take care of trailing spaces (this is the bare minimum), using built-in option of code editors (`"files.trimTrailingWhitespace": true` in VScode)

What do you prefer ?

**Comment by jejellyroll-fr** [comment]
hummmm ... i see ... for people like me who are using a lot of different vm ... until the project has reached a sufficient level of maturity i prefer to add on the project:

an editorconfig → sets the basics (indentation, EOL, newline) and a pre-commit → auto-cleans (spaces, EOL) + light Ruff lint. (E,F,I,UP,W)

then Normal Git commit

**Comment by jejellyroll-fr** [comment]
deliver in dev https://github.com/jejellyroll-fr/fpdb-3/issues/102
